storage/csv/datasource.go

```diff
@@ -112,13 +139,161 @@ func (rs *RecordStream) Close() error {
+			return
+		}
+
+		tx.Abort() // We only read data above, no need to risk failing now.
+
+		// Moving file iterator by `rs.offset`
+		for i := 0; i < rs.offset; i++ {
+			_, err := rs.readRecordFromFileWithInitialize()
+			if err == execution.ErrEndOfStream {
+				return
+			} else if err != nil {
+				log.Fatalf("csv worker: couldn't move csv file iterator by %d offset: %s", rs.offset, err)
+				return
+			}
+		}
+
+		for { // inner for is calling RunWorkerInternal
+			tx := rs.stateStorage.BeginTransaction().WithPrefix(rs.streamID.AsPrefix())
+
+			err := rs.RunWorkerInternal(ctx, tx)
+			if errors.Cause(err) == execution.ErrNewTransactionRequired {
+				tx.Abort()
+				continue
+			} else if waitableError := execution.GetErrWaitForChanges(err); waitableError != nil {
+				tx.Abort()
+				err = waitableError.ListenForChanges(ctx)
+				if err != nil {
+					log.Println("csv worker: couldn't listen for changes: ", err)
+				}
+				err = waitableError.Close()
+				if err != nil {
+					log.Println("csv worker: couldn't close storage changes subscription: ", err)
+				}
+				continue
+			} else if err == execution.ErrEndOfStream {
+				err = tx.Commit()
+				if err != nil {
+					log.Println("csv worker: couldn't commit transaction: ", err)
+					continue
+				}
+				return
+			} else if err != nil { // TODO - should this close worker? This could be a case where file is somehow faulty (question in PR)
+				tx.Abort()
+				log.Printf("csv worker: error running csv read batch worker: %s, reinitializing from storage", err)
+				break
+			}
+
+			err = tx.Commit()
+			if err != nil {
+				log.Println("csv worker: couldn't commit transaction: ", err)
+				continue
+			}
+		}
 	}
+}
 
-	return resultRow
+var outputQueuePrefix = []byte("$output_queue$")
+
+func (rs *RecordStream) RunWorkerInternal(ctx context.Context, tx storage.StateTransaction) error {
+	outputQueue := execution.NewOutputQueue(tx.WithPrefix(outputQueuePrefix))
+
+	if rs.isDone {
+		err := outputQueue.Push(ctx, &QueueElement{
+			Type: &QueueElement_Error{
+				Error: execution.ErrEndOfStream.Error(),
+			},
+		})
+		if err != nil {
+			return errors.Wrapf(err, "couldn't push csv EndOfStream to output record queue")
+		}
+
+		log.Println("csv worker: ErrEndOfStream pushed")
+		return execution.ErrEndOfStream
+	}
+
+	batch := make([]*execution.Record, 0)
+	for i := 0; i < rs.batchSize; i++ {
+		aliasedRecord, err := rs.readRecordFromFileWithInitialize()
+		if err == execution.ErrEndOfStream {
+			break
+		} else if err != nil { // TODO - do we want to return err instantly here or push current batch? (question in PR)
+			return errors.Wrap(err, "couldn't read record from csv file")
+		}
+
+		batch = append(batch, execution.NewRecord(
+			rs.aliasedFields,
+			aliasedRecord,
+			execution.WithID(execution.NewRecordIDFromStreamIDWithOffset(rs.streamID, rs.offset+i))))
+	}
+
+	for i := range batch {
+		err := outputQueue.Push(ctx, &QueueElement{
```

Comment:
This applies to all places this appears - I think it would be more optimal to write a `NewQueueRecordElement` or something of that sort and use it everywhere, because 4 lines of code is a bit of an overkill to create the queue element, and this appears more than once throughout the code

Reply:
I wrote about that in cube's kafka PR (I mean, exactly the same question) and we've come into conclusion that it wouldn't be necessary so I copy-pasted the way kafka datasource does it
@cube2222

Reply:
yea, don't abstract too early. I actually think that we'll be able to abstract out most of the "revival" logical in the future.